CIPs/0008.md

```diff
@@ -64,14 +63,56 @@ I considered doing storing all data on IPFS, however the tooling around IPFS as
+
+### Privacy
+
+We commonly expect users to share data only with a set of authorized peers. A profile picture, under the path /profile/picture might be shared only with the user’s friends. Creating and publishing a number of ciphertexts to a related path (e.g. /profile/picture/{recipient address}), each encrypted with a friend's public key would be one way to achieve this. It would ensure data integrity and confidentiality.
+
+A problem with indexing the ciphertexts by recipient is that it reveals publicly every recipient of a given message. In the case of profile pictures, this would make public every “friend” with whom the picture is shared, and in the case of payment requests it would reveal that a request is pending from the user to a given peer. This data may be valuable to attackers looking to craft a spear-phishing attack, and may violate some users’ expectations of privacy.
+
+To solve this we present a scheme that maintains both sender and receiver privacy, to compute the storage path it requires the publisher's private data encryption key and the receivers public data encryption key. The path can be computed as follows:
+
+```
+label = PRF(ECDH(A, B), A || B || data_path)
+ciphertext path = "/ciphertexts/" || hex(label)
+```
+
+Where A is the sender’s public key, B is the recipient’s public key, ECDH is a non-interactive key exchange requiring the private key for either A or B to evaluate, PRF is the pseudorandom function HMAC-BLAKE2s-256, "ciphertexts" is a cosmetic path URL path prefix, and data pathis the URL path suffix. ECDH produces a secp256k1 elliptic curve point, which is encoded and used as the 64B key for PRF. Elliptic curve points (e.g. public keys and shared secrets) are represented in SEC1-v2 compressed representation. The message input to the PRF consists of the concatenation of the public key of the sender and the public key of the recipient (both in SEC1-v2 compressed representation), and the UTF8 encoding of the data path.
+
+`A || B` is prepended to the data path for future compatibility with sender hiding. If both the sender and receiver upload messages to the same path prefix in order to share an “anonymity pool” and wish to, e.g., mutually share photos under the data path “profile/photo,” then this ensures there’s a unique ciphertext path for A sending a photo to B and B sending a photo to A.
+
+A ciphertext path is easily computable from the data path and public keys of the sender and recipient, and knowledge of either private key. It is infeasible to compute without access to one of the two private keys, and it is also infeasible to determine the sender or recipient. (Although in CIP8 the sender is known because storage roots are not shared.)
```

Comment:
```suggestion
A ciphertext path is easily computable from the data path and public keys of the sender and recipient, and knowledge of either private key. It is infeasible to compute without access to one of the two private keys, and it is also infeasible to determine the sender or recipient. (Although in this specification, the sender is known because storage roots are not shared.)
```